LLama/Native/SafeLlamaModelHandle.cs

```diff
@@ -576,13 +614,39 @@ public SafeLLamaContextHandle CreateContext(LLamaContextParams @params)
         /// <summary>
         /// Get tokens for a model
         /// </summary>
-        public class ModelTokens
+        public readonly struct ModelTokens
         {
             private readonly SafeLlamaModelHandle _model;
+            private readonly string? _eot;
+            private readonly string? _eos;
 
             internal ModelTokens(SafeLlamaModelHandle model)
             {
                 _model = model;
+                _eot = LLamaTokenToString(EOT, true);
+                _eos = LLamaTokenToString(EOS, true);
+            }
+
+            private string? LLamaTokenToString(LLamaToken? token, bool isSpecialToken)
```

Comment:
Handling strings in this class feels like a bit of a mismatch to me (it's meant to just e a simple list of tokens). As far as I can see these new string properties are only used up in the Antiprompts (which may not be necessary).

If they're not critical to anything else can we remove this change from this PR for now, and discuss if we want something like this and where best to put it in another PR?

Reply:
> Handling strings in this class feels like a bit of a mismatch to me (it's meant to just e a simple list of tokens). As far as I can see these new string properties are only used up in the Antiprompts (which may not be necessary).

This was my primary use-case, yes. Per your comment above, I need to explore the `IsEndOfGeneration`. Is the `InteractiveExecutor` sticking around? My motivation for adding this was because of the chat examples that used the interactive executor. I (naively) thought I could just plug in any model and it would work. I tried with Llama3 and Phi3 and got some odd behavior until I learned about chat templates. That made me think we could synthesize the history transformer to make a best guess and use the llama.cpp template formatter. To your other comments though, it's not great to have to constantly allocate memory to do the P/Invokes and I did consider if it's worth just porting the code to be dotnet native but it would stink to have to keep it in lockstep with that repo so I finally landed on the implementation in this PR. (this would have been helpful in the initial PR description 😅 )

To me, the `ModelToken` class seemed like a best fit and could cache the results to prevent further calls to the native library but I could remove it and just expose this function as a utility if that seems better

Reply:
> Is the InteractiveExecutor sticking around

Probably (hopefully) not. My long term project for the past 6 months or so has been to build a ew foundation for better future executors.

> I (naively) thought I could just plug in any model and it would work

In the long term, I hope we can make that work! Many of the bits are there now - Automatic antiprompting (with end-of-generation tokens), automatic templating (with LlamaTemplate), maybe even automatic context shifting to handle out-of-context errors!

> To me, the ModelToken class seemed like a best fit and could cache the results to prevent further calls to the native library but I could remove it and just expose this function as a utility if that seems better

I'm ok with either way to be honest.

If we're going to keep it with the strings in `ModelTokens` could you switch ModelToken back to a `sealed class` instead of a struct - it's a bit too big to be a struct (and will probably grow if we add equivalent stirng properties for all the special tokens.

Reply:
> Probably (hopefully) not. My long term project for the past 6 months or so has been to build a ew foundation for better future executors.

Is that going to be here or another project entirely? I'd be interested in approaching this from the ground up if you're open to it 😸 

I'll revert the `ModelTokens` back to a sealed class and leave it as-is for now if that's cool

Reply:
Here. Over the past six months I've been laying the groundwork with `BatchedExecutor` (a low level executor that can act as a foundation for all executors).

I wrote up a lot of my thoughts recent in this comment [this comment](https://github.com/SciSharp/LLamaSharp/pull/740#issuecomment-2117693956). I ended up giving an overview of how I see the project being split up into layers. In particular the **Mid level** and **Puzzle Pieces** section show how I've put together a lot of the parts for new high level executors, now just need to figure out how to put them together! I'd be very happy with help there!

Reply:
Oh cool, I'll give that thread a read to get caught up